File: Java/Classes/EuclideanVector.java
package Classes;

import java.awt.geom.Point2D;
import java.lang.*;

public class EuclideanVector extends BaseClass {
    private Point2D p1 = new Point2D.Double();
    private Point2D p2;
    private double length;
    private double angle;

    public EuclideanVector(double l, double a, Point2D p, String n)
    {
        super(n);

        if( Math.abs(a) == 90 || Math.abs(a) == 180)
        {
            p2 = new Point2D.Double(0,l);
        }
        else
            if (Math.abs(a) == 0 || Math.abs(a) == 360)
            {
                p2 = new Point2D.Double(l,0);
            }
            else
                p2 = new Point2D.Double(Math.cos(a * Math.PI / 180) * l, Math.sin(a * Math.PI / 180) * l);

        length = l;
        angle = a;
    }

    public EuclideanVector(double l, double a, String n)
    {
        super(n);
        p2 = new Point2D.Double( Math.cos(a*Math.PI/180)*l, Math.sin(a*Math.PI/180)*l );
        length = l;
        angle = a;
    }

    public EuclideanVector(String n, Point2D pa, Point2D pb)
    {
        super(n);
        p1 = pa;
        p2 = pb;
        length = Math.sqrt( Math.pow(p2.getX()-p1.getX(),2) + Math.pow(p2.getY()-p1.getY(),2) );
        angle = Math.atan ((p2.getY()-p1.getY()) / (p2.getX()-p1.getX()) ) * 180 / Math.PI;

    }

    public EuclideanVector(EuclideanVector v)
    {
        super(v.getName());
        p1=v.p1;
        p2=v.p2;
        length=v.length;
        angle=v.angle;
    }

    public EuclideanVector()
    {
        super("null");
        p1 = new Point2D.Double();
        p2 = new Point2D.Double();
        length = 0;
        angle = 0;
    }

    public double getA() { return angle; }

    public double getL() { return length; }

    public static double Angle(EuclideanVector v1, EuclideanVector v2) //angolo tra due vettori
    {
        return Math.acos( ( (v1.dotProduct(v2)) / (v1.length*v2.length) ) * Math.PI/180);
    }

    public EuclideanVector crossProduct(EuclideanVector v) //prodotto vettoriale
    {
        return  new EuclideanVector(v.length*length*Math.sin(Angle(this,v)),0,"Result");
    }

    public double dotProduct(EuclideanVector v) //prodotto scalare
    {
        return p2.getX()*v.p2.getX()+p2.getY()*v.p2.getY();
    }

    public EuclideanVector sum(EuclideanVector v)
    {
        Point2D p = new Point2D.Double(p1.getX()+p2.getX(),v.p1.getY()+v.p2.getY());
        return new EuclideanVector("result", p, new Point2D.Double());
    }

    public EuclideanVector minum(EuclideanVector v)
    {
        Point2D p = new Point2D.Double(p1.getX()-p2.getX(),v.p1.getY()-v.p2.getY());
        return new EuclideanVector("result", p, new Point2D.Double());
    }

    public EuclideanVector molt(double n)
    {
        return new EuclideanVector(length*=n,angle,getName());
    }

    public String print()
    {
        String aux="Vettore  " + super.print();
        aux += "\nPunto di applicazione : ("+ p1.getX() + ',' + p1.getY() + ")" + "\nPunto 2 : (" + p2.getX() + ',' + p2.getY() + ")";
        aux += "\nModulo = " + length;
        aux += "\nAngolo = " + angle;
        return aux;
    }
}

File: euclideanvector.cpp
#include "euclideanvector.h"
#include <typeinfo>
#include "exception.h"
#include <iostream>
#include <math.h>

EuclideanVector::EuclideanVector(double l, double a, QPointF p, QString n) : BaseClass (n),
  p1(p), p2(QPointF(l*cos(a*M_PI/180),l*sin(a*M_PI/180))), length(l), angle(a)
{
  if(std::abs(a)==90.0 || std::abs(a)==180.0)
    p2=QPointF(0,l);
  else if (std::abs(a)==0.0 || std::abs(a)==360.0)
    {
      p2=QPointF(l,0);
    }
}

//costruisce un vettore che ha come punto di applicazione l'origine
EuclideanVector::EuclideanVector(double l, double a, QString n) : BaseClass (n),
  p1(QPointF()), p2(QPointF(l*cos(a*M_PI/180),l*sin(a*M_PI/180))), length(l), angle(a)
{

  if(std::abs(a)==90.0 || std::abs(a)==180.0)
    p2=QPointF(0,l);
  else if (std::abs(a)==0.0 || std::abs(a)==360.0)
    {
      p2=QPointF(l,0);
    }
}

EuclideanVector::EuclideanVector(QString n, QPointF p2, QPointF p1) : BaseClass (n), p1(p1), p2(p2),
  length(sqrt(pow((p2.x()-p1.x()),2)+pow((p2.y()-p1.y()),2))), angle(atan((p2.y()-p1.y())/(p2.x()-p1.x()))*180/M_PI)
{}

EuclideanVector::EuclideanVector(double a, QString n) : BaseClass(n), p1(QPointF(0,0)), p2(QPointF(0,0)),
  length(a), angle(0) {}

EuclideanVector::EuclideanVector(const EuclideanVector& v) : BaseClass (v.getName()), p1(v.p1), p2(v.p2), length(v.length), angle(v.angle) {}

EuclideanVector::EuclideanVector() : BaseClass(" "), p1(), p2(), length(0), angle(0) {}

double EuclideanVector::dotProduct(const EuclideanVector& v) const
{
  return p2.x()*v.p2.x()+p2.y()*v.p2.y();
}

//angolo tra due vettori
double EuclideanVector::Angle(const EuclideanVector &v1, const EuclideanVector &v2)
{
  return acos( ( (v1.dotProduct(v2)) / (v1.length*v2.length) ) *M_PI/180);
}

EuclideanVector* EuclideanVector::operator+(const BaseClass &v) const
{
  if(typeid(*this) != typeid(v))
    throw InvalidOperation();

  QPointF p(p1.x()+p2.x(),static_cast<const EuclideanVector&>(v).p1.y()+static_cast<const EuclideanVector&>(v).p2.y());
  return new EuclideanVector("result",p);
}

EuclideanVector* EuclideanVector::operator-(const BaseClass &v) const
{
  if(typeid(*this) != typeid(v))
    throw InvalidOperation();

  QPointF p(p1.x()-p2.x(),static_cast<const EuclideanVector&>(v).p1.y()-static_cast<const EuclideanVector&>(v).p2.y());
  return new EuclideanVector("result",p);
}


EuclideanVector* EuclideanVector::operator*(const BaseClass &v) const //prodotto scalare
{
  if(typeid(*this) != typeid(v))
    throw InvalidOperation();

  return new EuclideanVector(this->dotProduct(static_cast<const EuclideanVector&>(v)),"DotProduct");
}

EuclideanVector* EuclideanVector::operator/(const BaseClass &v) //prodotto vettoriale
{
  if(typeid(*this) != typeid(v))
    throw InvalidOperation();

  delete this;
  *this = EuclideanVector(length * static_cast<const EuclideanVector&>(v).length *
                             sin(Angle(*this,static_cast<const EuclideanVector&>(v))*M_PI/180),0,"Result");
  return this;
}

QString EuclideanVector::print() const
{
  if(angle==0. && p1==QPointF(0,0) && p2==QPointF(0,0))
    return QString("Prodotto Scalare = "+QString::number(length));

  QString aux="Vettore  "+BaseClass::print();
  aux.append("\r\nPunto di applicazione : ("+QString::number(p1.x())+','+QString::number(p1.y())+")" +
             "   Punto 2 : ("+QString::number(p2.x())+','+QString::number(p2.y())+")");
  aux.append("\r\nModulo = "+QString::number(length));
  aux.append("\r\nAngolo = "+QString::number(angle));
  return aux;
}


